app/src/main/java/lu/geoportail/map/LuxTileServer.kt

```diff
@@ -35,71 +47,241 @@ class LuxTileServer(
         } catch (e: IOException) {
             Log.e("tag", "Failed to get asset file list.", e)
         }
-        if(!toPathRoot.exists()) Files.createDirectory(toPathRoot.toPath())
+        if(!toPathRoot.exists()) Files.createDirectories(toPathRoot.toPath())
         if (files != null) for (filename in files) {
+            val sourceFileBytes = try {
+                assetManager.open("$assetPath/$filename").use { it.readBytes() }
+            }
+            catch (e: IOException) {
+                // exception caught if filename designates a subfolder instead of a file
+                continue
+            }
             val file = File(toPathRoot, filename)
             if (file.exists()) continue
 
-            val sourceFileBytes = assetManager.open("$assetPath/$filename").readBytes()
             FileOutputStream(file).use {
                 it.write(sourceFileBytes)
             }
         }
     }
 
+    private fun downloadAssetsFromMeta(json: JSONObject, toPathRoot: File) {
+        val sources = json.getJSONArray("sources")
+        for (i in 0 until sources.length()) {
+            downloadAssets(sources.getString(i), toPathRoot)
+        }
+    }
+
+    private fun downloadAssets(assetUrl: String, toPathRoot: File) {
+        val url = URL(assetUrl)
+        val filename =url.file
+        val file = File(toPathRoot, filename)
+        if(!file.parentFile.exists()) Files.createDirectories(file.parentFile.toPath())
+
+        val sourceStream = BufferedInputStream(url.openStream())
+        val fileOutputStream = FileOutputStream(file)
+        val buf = ByteArray(1024)
+        var len: Int
+        while (sourceStream.read(buf).also { len = it } > 0) {
+            fileOutputStream.write(buf, 0, len)
+        }
+        sourceStream.close()
+        fileOutputStream.close()
```

Comment:
No need see my "Kolinize the code PR" #4

Reply:
unfortunately reading the complete file to memory crashed the app because the temp buffer exceeded heap size on big files.
so I had to revert to the previous version.

Reply:
I might try again with the BufferedInputStream giving it a max size. (`BufferedInputStream(InputStream in, int size`)
)